### PR TITLE
Add save_state_dict and load_state_dict functions (fixes #150857)

### LINKED ISSUE #150857: No way to save/load model weights between PyTorch and LibTorch (C++) in a compatible way
### 🚀 The feature, motivation and pitch

LibTorch (C++) does not provide a way to save or load model weights in a way compatible with Python’s torch.save() / torch.load() or state_dict() / load_state_dict(). This makes it hard to share models between Python and C++ environments.

### Alternatives

I've managed to implement cross-compatibility between PyTorch (Python) and LibTorch (C++) by creating my own binary weight serializer. Currently, LibTorch lacks a native and simple way to save/load only model weights that are compatible with Python torch.nn.Module.state_dict() and vice versa.

Here is a working example of what I had to do manually (binary format):
**C++ Code (LibTorch)**
```cpp
void save_weights(const std::string& path) {
    std::ofstream file(path, std::ios::binary);
    to(torch::kCPU);
    for (const auto& param : parameters()) {
        auto data = param.data().contiguous();
        auto size = data.numel();
        file.write(reinterpret_cast<const char*>(&size), sizeof(size));
        file.write(reinterpret_cast<const char*>(data.data_ptr<float>()), size * sizeof(float));
    }
    file.close();
}

void load_weights(const std::string& path) {
    std::ifstream file(path, std::ios::binary);
    for (auto& param : parameters()) {
        int64_t size;
        file.read(reinterpret_cast<char*>(&size), sizeof(size));
        std::vector<float> param_data(size);
        file.read(reinterpret_cast<char*>(param_data.data()), size * sizeof(float));
        param.data().copy_(torch::tensor(param_data).view_as(param));
    }
    file.close();
}
```
**Python Code**
```python
def save_binary_weights(self, path):
    with open(path, "wb") as file:
        for param in self.parameters():
            data = param.data.contiguous().cpu().view(-1).numpy().astype("float32")
            size = len(data)
            file.write(struct.pack('Q', size))
            file.write(struct.pack(f'{size}f', *data))

def load_binary_weights(self, path: str):
    with open(path, "rb") as file:
        for param in self.parameters():
            size = struct.unpack('Q', file.read(8))[0]
            param_data = struct.unpack(f'{size}f', file.read(size * 4))
            param.data.copy_(torch.tensor(param_data).view_as(param))
```
**Why this matters**
LibTorch doesn’t support Python’s torch.save() and torch.load() directly due to incompatible serialization mechanisms. Also, TorchScript lacks the ability to export model weights easily for reloading in C++.

**It would be very helpful if PyTorch could support either of the following:**

- A standard binary format for model weights compatible between Python ↔ C++.
- A .load_state_dict() and .state_dict() equivalent in LibTorch.
- Built-in utility to read .pt or .pth weight-only files in C++.

This would massively simplify workflows for deploying or testing models across environments.

### Additional context

You see full code in the my repo [Toguzkumalak](https://github.com/FuryBaM/Toguzkumalak/tree/master/Toguzkumalak).
Results in libtorch c++
```
Policy: [0.110181, 0.122630, 0.111052, 0.103369, 0.105678, 0.114038, 0.108151, 0.115452, 0.109449]
Value prediction: 0.051766
State: [9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 0.000000, 0.000000, 0.000000]
```

in python
```python
if __name__ == "__main__":
    model = TNET()
    model.load_binary_weights("C:/Users/Akzhol/source/repos/Toguzkumalak/Toguzkumalak/build/Release/model_data/weights.dat")
    input_data = torch.tensor([9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 
                            9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 9.000000, 
                            9.000000, 9.000000, 0.000000, 0.000000, 0.000000], dtype=torch.float32).view(1, -1)
    model.eval()
    output = model(input_data)
    print(output)
```
```
(tensor([[0.1102, 0.1226, 0.1111, 0.1034, 0.1057, 0.1140, 0.1082, 0.1155, 0.1094]],
       grad_fn=<ExpBackward0>), tensor([[0.0518]], grad_fn=<TanhBackward0>))
```

### PR DESCRIPTION
**Pull request by chani0343**
## Summary
This PR adds `torch::save_state_dict()` and `torch::load_state_dict()` template functions to enable C++ compatibility with Python's `torch.save(model.state_dict())` format.

## Fixes
Closes #150857

## Problem
Currently, there's no direct way in C++ PyTorch to save/load only the parameters and buffers of a module (state dictionary) in a format compatible with Python's `torch.save(model.state_dict())`. The existing `torch::save()` and `torch::load()` functions serialize the entire module structure, not just the state dictionary.

## Solution
- **`save_state_dict()`**: Saves only module parameters and buffers using `OutputArchive`
- **`load_state_dict()`**: Loads parameters and buffers into an existing module using `InputArchive`
- Uses `NoGradGuard` to prevent gradient computation during loading
- Compatible with existing PyTorch serialization system

## Changes
- Added `save_state_dict()` template function in `torch/csrc/api/include/torch/serialize.h`
- Added `load_state_dict()` template function in `torch/csrc/api/include/torch/serialize.h`
- Added comprehensive test case `TEST(SerializeTest, StateDict)` in `test/cpp/api/serialize.cpp`
- Removed verbose documentation comments to match existing code style

## Testing
- Unit test verifies save/load functionality with `torch::nn::Linear` model
- Tested parameter and buffer serialization/deserialization
- Confirmed weights and biases are correctly preserved across save/load cycle
- Manual testing confirmed compatibility with existing serialization system

## Usage Example
```cpp
auto model1 = torch::nn::Linear(3, 2);
auto model2 = torch::nn::Linear(3, 2);

// Save state dict
torch::save_state_dict(model1, "model_state.pt");

// Load state dict
torch::load_state_dict(model2, "model_state.pt");
```
This enables seamless interoperability between Python and C++ PyTorch for state dictionary operations.

### CODE REVIEW

**Comment by chani0343** [comment]
@pytorchbot label "no-stale"

@jbschlosser @mikaylagawarecki 

This PR adds essential C++ API functions for state dictionary operations, enabling compatibility between Python and C++ PyTorch (fixes #150857).

Key features:
✅ torch::save_state_dict() and torch::load_state_dict() functions
✅ Compatible with Python's torch.save(model.state_dict()) format  
✅ Comprehensive test coverage
✅ All checks passing
✅ Addresses real user need for C++/Python interoperability

@jerryzh168 requested your review in March. Could you please take a look when you have a moment?

Thanks!
